clients/adapter/src/main/java/org/eclipse/hono/adapter/client/command/CommandRouterCommandConsumerFactory.java

```diff
@@ -11,100 +11,100 @@
      * @param commandRouterClient The client to use for accessing the command router service.
+     * @param adapterName The name of the protocol adapter.
      * @throws NullPointerException if any of the parameters is {@code null}.
      */
-    public ProtonBasedCommandRouterCommandConsumerFactoryImpl(
-            final HonoConnection connection,
-            final SendMessageSampler.Factory samplerFactory,
-            final CommandRouterClient commandRouterClient) {
-        super(connection, samplerFactory);
+    public CommandRouterCommandConsumerFactory(final CommandRouterClient commandRouterClient, final String adapterName) {
         this.commandRouterClient = Objects.requireNonNull(commandRouterClient);
-
-        adapterInstanceId = getAdapterInstanceId(connection.getConfig().getName());
-        adapterInstanceCommandHandler = new ProtonBasedAdapterInstanceCommandHandler(connection.getTracer(), adapterInstanceId);
+        this.adapterInstanceId = getNewAdapterInstanceId(Objects.requireNonNull(adapterName));
     }
 
-    private static String getAdapterInstanceId(final String adapterName) {
+    private static String getNewAdapterInstanceId(final String adapterName) {
         // replace special characters so that the id can be used in a Kafka topic name
         final String prefix = Strings.isNullOrEmpty(adapterName) ? ""
                 : adapterName.replaceAll("[^a-zA-Z0-9._-]", "") + "-";
         return prefix + UUID.randomUUID();
     }
 
+    /**
+     * Registers the command consumer receiving commands on the internal command endpoint of the protocol adapter.
+     * That is the endpoint that the Command Router forwards received commands to.
+     *
+     * @param internalCommandConsumerSupplier Function that returns the consumer. Parameters are the adapter instance
+     *            id that identifies the internal command endpoint and the command handlers to choose from when handling
+     *            a received command.
+     */
+    public void registerInternalCommandConsumer(
+            final BiFunction<String, CommandHandlers, Lifecycle> internalCommandConsumerSupplier) {
+        internalCommandConsumers.add(internalCommandConsumerSupplier.apply(adapterInstanceId, commandHandlers));
+    }
+
     @Override
     public Future<Void> start() {
-        return super.start()
-                .onComplete(v -> {
-                    connection.addReconnectListener(c -> recreateConsumer());
-                    // trigger creation of adapter specific consumer link (with retry if failed)
-                    recreateConsumer();
-                });
+        @SuppressWarnings("rawtypes")
+        final List<Future> futures = internalCommandConsumers.stream()
+                .map(Lifecycle::start)
+                .collect(Collectors.toList());
+        if (futures.isEmpty()) {
+            return Future.failedFuture("no command consumer registered");
+        }
+        return CompositeFuture.all(futures).mapEmpty();
```

Comment:
I do not understand this start-up behavior. When is the star method being invoked and by whom? Why should there be any *internalCommandConsumers* being (already) registered at this point? Why should start-up fail if there aren't any? IMHO all of this deserves proper documentation in the method's JavaDoc as it is far from obvious what this method is supposed to do ...

Reply:
The `CommandRouterCommandConsumerFactory` instance (or a `ProtonBasedDelegatingCommandConsumerFactory` instance) is kept in the `AbstractProtocolAdapterBase#commandConsumerFactory` field. 
In `AbstractProtocolAdapterBase`, the `start` method is invoked on that `commandConsumerFactory`.

The `CommandRouterCommandConsumerFactory` needs the (started) _InternalCommandConsumer(s)_ (proton and/or kafka based) to work. Therefore I've associated the startup of the consumer factory with the startup of the _InternalCommandConsumer(s)_. Otherwise, the `AbstractProtocolAdapterBase` class would have to deal with the startup of the _InternalCommandConsumer(s)_ explicitly (and only do so if a `CommandRouterCommandConsumerFactory`, not a `ProtonBasedDelegatingCommandConsumerFactory` is set).

I've adapted the javadocs of the `registerInternalCommandConsumer` method and the above `start` method to describe more clearly what's done and required.